# Wrangle

In [1]:
import pandas as pd

**Data**

In [2]:
migration = pd.read_excel("https://query.data.world/s/al6wfboprlcexlbsk5a44el7ukdqhr?dws=00000")
migration.head()

Country STATE OR PROVINCE  Year  Inbound  Outbound
0  United States           ALABAMA  2022     0.57      0.43
1  United States           ALABAMA  2021     0.59      0.41
2  United States           ALABAMA  2020     0.61      0.39
3  United States           ALABAMA  2019     0.57      0.43
4  United States           ALABAMA  2018     0.56      0.44

In [3]:
migration.describe()

Year     Inbound    Outbound
count   610.000000  610.000000  610.000000
mean   2017.500000    0.477689    0.510869
std       2.874639    0.100194    0.102179
min    2013.000000    0.000000    0.000000
25%    2015.000000    0.420000    0.450000
50%    2017.500000    0.490000    0.510000
75%    2020.000000    0.540000    0.570000
max    2022.000000    0.750000    1.000000

In [4]:
migration.shape

(610, 5)

In [5]:
migration.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 610 entries, 0 to 609
Data columns (total 5 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Country            610 non-null    object 
 1   STATE OR PROVINCE  610 non-null    object 
 2   Year               610 non-null    int64  
 3   Inbound            610 non-null    float64
 4   Outbound           610 non-null    float64
dtypes: float64(2), int64(1), object(2)
memory usage: 24.0+ KB


In [6]:
# make columns lower case
migration.columns = migration.columns.str.lower().str.replace(" ", "_")
migration.columns

Index(['country', 'state_or_province', 'year', 'inbound', 'outbound'], dtype='object')

In [7]:
# check for null values
migration.isna().sum()

country              0
state_or_province    0
year                 0
inbound              0
outbound             0
dtype: int64

In [8]:
# make year categorical
migration.year = migration.year.astype("category")

In [9]:
migration.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 610 entries, 0 to 609
Data columns (total 5 columns):
 #   Column             Non-Null Count  Dtype   
---  ------             --------------  -----   
 0   country            610 non-null    object  
 1   state_or_province  610 non-null    object  
 2   year               610 non-null    category
 3   inbound            610 non-null    float64 
 4   outbound           610 non-null    float64 
dtypes: category(1), float64(2), object(2)
memory usage: 20.2+ KB


In [11]:
# save data
migration.to_csv("Household_goods_migration_patterns.csv")

**Take aways**

- 610 rows and 5 columns
- 0 null counts
- values contained in in/out bound are total shipment percetange

In [30]:
import tabula
import pandas as pd

# Specify the path to your PDF file
pdf_path = "2022_Migration_Patterns PDF MAP.pdf"

# Extract tables from the PDF and store them in a list of DataFrames
tables = tabula.read_pdf(pdf_path, pages=1, multiple_tables=True)

# Loop through the list of DataFrames and process the tables as needed
for i, table in enumerate(tables):
    # Process each table, for example, print it
    print(f"Table {i + 1}:")
    t = table


Table 1:


In [31]:
t

Unnamed: 0  56% Unnamed: 1   ID  Unnamed: 2 Unnamed: 3  Unnamed: 4  \
0         NaN  44%        NaN  49%         NaN        NaN         NaN   
1         NaN  NaN        NaN  51%         NaN         WY         NaN   
2         NaN  NaN        NaN  NaN         NaN    58% 42%         NaN   
3         NaN  NaN         NV  NaN         NaN        NaN         NaN   
4         NaN  NaN        49%   UT         NaN        NaN         NaN   
5         NaN  NaN        51%  45%         NaN         CO         NaN   
6         NaN   CA        NaN  55%         NaN        47%         NaN   
7         NaN  52%        NaN  NaN         NaN        53%         NaN   
8         NaN  48%        NaN  NaN         NaN        NaN         NaN   
9         NaN  NaN        NaN  NaN         NaN        NaN         NaN   
10        NaN  NaN        NaN   AZ         NaN         NM         NaN   
11        NaN  NaN        NaN  51%         NaN     41%42%         NaN   
12        NaN  NaN        NaN  49%         NaN     59%58%         NaN   
13        NaN  NaN        NaN  NaN         NaN        NaN         NaN   
14        NaN  NaN        NaN  NaN         NaN        NaN         NaN   
15         HI  NaN        NaN  NaN         NaN        NaN         NaN   

   Unnamed: 5      42% WI  Unnamed: 6            41%   MA Unnamed: 7  \
0     CT51%SD         52%          MI            NaN  NaN        NaN   
1         49%         48%         58%            NaN  NaN        NaN   
2          NE      IA 58%         42%  58% PA OH 42%  N J         RI   
3         54%         42%       IL IN            55%   DE        NaN   
4         46%         61%         59%      45% WV VA  NaN        NaN   
5         NaN      39% MO         41%     47%53% 45%   MD        NaN   
6         NaN  KS 53% 51%         NaN        47% 55%   DC        NaN   
7         NaN     47% 49%          KY        53% 36%  NaN        NaN   
8         NaN         NaN         NaN     NC 42% 64%  NaN        NaN   
9         NaN          OK          TN         58% SC  NaN        NaN   
10        NaN      51% AR         NaN            51%  NaN        NaN   
11        NaN         49%          AL         GA 49%  NaN        NaN   
12        NaN         NaN          MS            NaN  NaN        NaN   
13        NaN         NaN     57% 43%            45%  NaN        NaN   
14        NaN         NaN  48%43% 57%            55%  NaN        NaN   
15        NaN          LA         NaN            NaN  NaN        NaN   

   Unnamed: 8 Unnamed: 9  
0         NaN        NaN  
1         NaN         NS  
2         NaN     41%59%  
3         NaN        NaN  
4          CT         DE  
5     53% 47%     54%46%  
6         NaN        NaN  
7          DC         MD  
8     44% 56%     51%49%  
9          MA         NH  
10        54%        43%  
11        46%        57%  
12        NaN        NaN  
13         NJ         RI  
14        55%        NaN  
15        45%        52%